x/protorev/keeper/hooks_test.go

```diff
@@ -0,0 +1,775 @@
+						TokenOut: "Atom",
+					},
+				},
+				executeSwap: func() {
+					_, err := s.App.PoolManagerKeeper.SwapExactAmountIn(s.Ctx, s.TestAccs[0], 1, sdk.NewCoin("akash", sdk.NewInt(100)), "Atom", sdk.NewInt(1))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "swap route exact amount in",
+			param: param{
+				expectedTrades: []types.Trade{
+					{
+						Pool:     1,
+						TokenIn:  "akash",
+						TokenOut: "Atom",
+					},
+				},
+				executeSwap: func() {
+
+					route := []poolmanagertypes.SwapAmountInRoute{{PoolId: 1, TokenOutDenom: "Atom"}}
+
+					_, err := s.App.PoolManagerKeeper.RouteExactAmountIn(s.Ctx, s.TestAccs[0], route, sdk.NewCoin("akash", sdk.NewInt(100)), sdk.NewInt(1))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "swap route exact amount out",
+			param: param{
+				expectedTrades: []types.Trade{
+					{
+						Pool:     1,
+						TokenIn:  "akash",
+						TokenOut: "Atom",
+					},
+				},
+				executeSwap: func() {
+
+					route := []poolmanagertypes.SwapAmountOutRoute{{PoolId: 1, TokenInDenom: "akash"}}
+
+					_, err := s.App.PoolManagerKeeper.RouteExactAmountOut(s.Ctx, s.TestAccs[0], route, sdk.NewInt(10000), sdk.NewCoin("Atom", sdk.NewInt(100)))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "swap route exact amount in - 2 routes",
+			param: param{
+				expectedTrades: []types.Trade{
+					{
+						Pool:     1,
+						TokenIn:  "akash",
+						TokenOut: "Atom",
+					},
+					{
+						Pool:     1,
+						TokenIn:  "Atom",
+						TokenOut: "akash",
+					},
+				},
+				executeSwap: func() {
+
+					route := []poolmanagertypes.SwapAmountInRoute{{PoolId: 1, TokenOutDenom: "Atom"}, {PoolId: 1, TokenOutDenom: "akash"}}
+
+					_, err := s.App.PoolManagerKeeper.RouteExactAmountIn(s.Ctx, s.TestAccs[0], route, sdk.NewCoin("akash", sdk.NewInt(100)), sdk.NewInt(1))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "swap route exact amount in - Concentrated Liquidity",
+			param: param{
+				expectedTrades: []types.Trade{
+					{
+						Pool:     49,
+						TokenIn:  "uosmo",
+						TokenOut: "epochTwo",
+					},
+				},
+				executeSwap: func() {
+
+					route := []poolmanagertypes.SwapAmountInRoute{{PoolId: 49, TokenOutDenom: "epochTwo"}}
+
+					_, err := s.App.PoolManagerKeeper.RouteExactAmountIn(s.Ctx, s.TestAccs[0], route, sdk.NewCoin("uosmo", sdk.NewInt(10)), sdk.NewInt(1))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+	}
+
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			s.SetupTest()
+			tc.param.executeSwap()
+
+			routes, err := s.App.ProtoRevKeeper.GetSwapsToBackrun(s.Ctx)
+			s.Require().NoError(err)
+			s.Require().Equal(tc.param.expectedTrades, routes.Trades)
+		})
+	}
+}
+
+// Tests the hook implementation that is called after liquidity providing
+func (s *KeeperTestSuite) TestLiquidityChanging() {
```

Comment:
q: we don't have CL liquidity changing infra because it does not affect price so nothing to backrun. Is my understanding correct?

Reply:
Yes, we only support backrunning GAMM exit/join because there are helper functions to make single-sided joins/exits, which swap in the background (and what we are really after is backrunning the swap).

To my knowledge, CL does not have a single-sided join/exit equivalent (that swaps -> provides liquidity), and thus I did not add any backrunning logic here.